src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs

```diff
@@ -1263,111 +1263,143 @@ public SolutionState WithDocumentFilePath(DocumentId documentId, string? filePat
             return UpdateDocumentState(oldDocument.UpdateFilePath(filePath));
         }
 
+        private SolutionState UpdateRelatedDocuments<TArg, TTextDocumentState>(
+            ImmutableArray<DocumentId> relatedDocuments,
+            Func<SolutionState, DocumentId, TTextDocumentState> getOldDocument,
+            Func<TTextDocumentState, TArg, bool> isSame,
+            Func<SolutionState, TTextDocumentState, TArg, SolutionState> updateSolution,
+            TArg arg) where TTextDocumentState : TextDocumentState
+        {
+            var newState = this;
+            foreach (var documentId in relatedDocuments)
+            {
+                var oldDocument = getOldDocument(newState, documentId);
+                if (isSame(oldDocument, arg))
+                    continue;
+
+                newState = updateSolution(newState, oldDocument, arg);
+            }
+
+            return newState;
+        }
+
         /// <summary>
         /// Creates a new solution instance with the document specified updated to have the text
         /// specified.
         /// </summary>
-        public SolutionState WithDocumentText(DocumentId documentId, SourceText text, PreservationMode mode = PreservationMode.PreserveValue)
+        public SolutionState WithDocumentText(
+            ImmutableArray<DocumentId> relatedDocuments,
+            SourceText text,
+            PreservationMode mode)
         {
-            var oldDocument = GetRequiredDocumentState(documentId);
-            if (oldDocument.TryGetText(out var oldText) && text == oldText)
-            {
-                return this;
-            }
-
-            return UpdateDocumentState(oldDocument.UpdateText(text, mode), textChanged: true);
+            return this.UpdateRelatedDocuments(
+                relatedDocuments,
+                static (newState, documentId) => newState.GetRequiredDocumentState(documentId),
```

Comment:
>static (newState, documentId) => newState.GetRequiredDocumentState(documentId) [](http://example.com/codeflow?start=16&length=78)

doesn't passing method group `GetRequiredDocumentState` instead of lambda work?

Reply:
no.  one is a delegate that takes two parameters, and the other is a method htat takes one.

Reply:
Ah, C# doesn't support this while the runtime does :(